### ISSUE TITLE
Unexpected behavior from Merge and Contour Tree (FTM)

### ISSUE THREAD

**Issue by ngorski9** [topology-tool-kit/ttk#990]
**Describe the bug**
I apologize if this is not an actual bug, but rather a misunderstanding. I have found a possibly unexpected behavior with the merge and contour tree filter. Namely, it appears that not every interior critical point in a dataset that is picked up by the ScalarFieldCriticalPoints filter is present as a node in the contour tree. My understanding is that every critical point in the interior of a data set should be a vertex of the contour tree.

**To Reproduce**
Note: All of this is is in Paraview.

1. Open the at.vti file that is a part of the ttk sample data. Make sure that the outline of this dataset is visible.
2. Apply a tetrahedralize filter. Uncheck the visibility of this filter.
3. Apply a ScalarFieldCriticalPoints filter to the tetrahedralize filter. Optionally, for visibility, apply an IcospheresFromPoints filter to the critical points filter.
4. Apply a Merge and Contour Tree (FTM) filter to the tetrahedralize filter. Uncheck the visibility of the Segmentation output channel

**Expected behavior**
Every interior critical point detected from the critical points filter should also be one of the vertices of the contour tree.

**Screenshots**
This image shows features from the at.vti dataset. The spheres shown are the critical points of the dataset, colored by critical type. The tubes are the branches of the contour tree. Some interior critical points that are not connected to the contour tree are circled.

![at-interior](https://github.com/topology-tool-kit/ttk/assets/62208712/352b4c60-9b20-492f-83af-638811b75eff)

Here is a synthetic example that I was able to create. I have a conjectured explanation for why this occurs for this specific dataset detailed in the "additional context" section. An interior critical point that is not connected to the contour tree is circled.

![interesting example](https://github.com/topology-tool-kit/ttk/assets/62208712/4b008483-fd8e-4eb7-a678-b49b935f8213)

**System (please complete the following information):**
 - OS: Ubuntu
 - Compiler: I'm not sure. I used the Ubuntu binary packages
 - Installed TTK's optional dependencies: N/A

**Additional context**

Here is the VTI file for the synthetic data that I used: [output.zip](https://github.com/topology-tool-kit/ttk/files/13419942/output.zip). This dataset is a 4x4x4 rectilinear grid. The value of most points in the dataset are assigned according to the formula 0.000001 * (3^x) * (5^y) * (7^z). The only exceptions to this are the labeled points below:

![interesting example labeled](https://github.com/topology-tool-kit/ttk/assets/62208712/b2f4dd7e-7f1a-48c3-a09d-4ad9b6678fff)

My conjecture is that this is a result of a specific step of the merge tree algorithm implemented in TTK [1].

The algorithm constructs the join and split trees, and then combines them to form the contour tree. My current understanding of how the join tree is calculated is that there are four steps:

1. A modified breadth first search is used to connect local minima to their respective saddle points. Points discovered along the way are identified with their respective branch.

2. A modified version of step (1) is used to connect the saddles from step (1) to new saddles. Specifically, this process begins at saddles whose entire lower link has already been visited. It repeats until there are no saddles remaining whose entire lower link has been visited. Points discovered along the way are identified with their respective branch.

3. Any remaining identified saddles whose entire lower link has not yet been visited are connected in ascending order of function value. Any remaining points that have not yet been visited are marked as ordinary points, and their segmentation ID is assigned to a branch created in this step according to their function value. That is, each remaining unclassified point x is assigned to the unique branch (a,b) that was created in this step such that f(a) < f(x) < f(b).

4. The minimum is connected to the join tree.

I conjecture that step 3 is the cause this behavior. Namely, it could be that there are some yet-undiscovered saddles that are supposed to connect to the remaining saddles. However these yet-undiscovered saddles could be skipped over when the extant saddles are automatically connected to one another.

In the synthetic example, this is perhaps demonstrated in the split tree, which has the same structure as the contour tree. All three of the saddles are connected to local maxima and must have therefore been discovered in step 1. After step 1, none of the present saddles have had their entire lower link visited, so step 2 is skipped. Then, in step 3, the saddles that were already found are connected to each other in order of function value, and the other points are assigned to the appropriate segmentation regions according to the algorithm. However, this misses the interior 1-saddle with function value 1.0. It should be noted that the join tree only has one branch that connects the maximum to the minimum. Thus, this interior saddle is missed in the calculation of both the join tree and the split tree.

It could be that this occurs due to the low resolution of this dataset. When analyzing the interior critical points of the at dataset that were not part of the contour tree, it appeared that they were all part of very small segmentation regions, which means that there could be a resolution issue occurring there as well.

Citation:

[1] Gueunet, Charles, Pierre Fortin, Julien Jomier, and Julien Tierny. "Task-based augmented merge trees with fibonacci heaps." In 2017 IEEE 7th Symposium on Large Data Analysis and Visualization (LDAV), pp. 6-15. IEEE, 2017.

**Comment by julien-tierny** [comment]
Hi, 
thanks for your message.

I do confirm that not all the critical points of a scalar field are captured by the Reeb graph (or the contour tree and its variants, including the merge tree). By definition, the Reeb graph is obtained by contracting connected components of level sets to points. Thus, branching (i.e. vertices) in the Reeb graph only occurs when the number of connected components of level sets change. 

However, it is possible to cross a critical value without changing the number of connected components. For example, in 3D, a level set component can "loop" back to itself and change its genus when passing a 1-saddle. In this example, the number of connected components does not change and the said saddle will therefore not be captured by the Reeb graph. 

The original paper by G. Reeb (in French) describes all the possible configurations. You can find a translation of this result in my habilitation thesis (page 55, property 12): https://julien-tierny.github.io/stuff/papers/tierny_habilitation.pdf

I hope this helps.

Best regards,
